mllib/src/main/scala/org/apache/spark/mllib/regression/GeneralizedLinearAlgorithm.scala

```diff
@@ -179,7 +179,7 @@ abstract class GeneralizedLinearAlgorithm[M <: GeneralizedLinearModel]
    */
   def run(input: RDD[LabeledPoint]): M = {
     if (numFeatures < 0) {
-      numFeatures = input.map(_.features.size).first()
+      numFeatures = input.first().features.size
```

Comment:
You could double-check with a very quick/small benchmark to see if it is in fact evaluated that way and if there is much difference at all. I suspect both are OK and don't know which is better.